File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans=max(nums)
        curr_max=1
        curr_min=1
        for i in nums:
            temp=curr_max*i
            curr_max=max(temp,curr_min*i,i)
            curr_min=min(temp,curr_min*i,i)
            ans=max(ans,curr_max)
        return ans    

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int pre=1,suf=1,n=nums.length,ans=Integer.MIN_VALUE;
        for(int i=0;i<n;i++){
            if(pre==0)pre=1;
            if(suf==0)suf=1;
            pre*=nums[i];
            suf*=nums[n-i-1];
            ans=Math.max(ans,Math.max(pre,suf));
        }
        return ans;
    }
}